agent/consul/session_timers.go

```diff
@@ -0,0 +1,82 @@
+package consul
+
+import (
+	"sync"
+	"time"
+)
+
+// SessionTimers provides a map of named timers which
+// is safe for concurrent use.
+type SessionTimers struct {
+	sync.RWMutex
+	m map[string]*time.Timer
+}
+
+func NewSessionTimers() *SessionTimers {
+	return &SessionTimers{m: make(map[string]*time.Timer)}
+}
+
+// Get returns the timer with the given id or nil.
+func (t *SessionTimers) Get(id string) *time.Timer {
+	t.RLock()
+	defer t.RUnlock()
+	return t.m[id]
+}
+
+// Set stores the timer under given id. If tm is nil the timer
+// witht the given id is removed.
+func (t *SessionTimers) Set(id string, tm *time.Timer) {
+	t.Lock()
+	defer t.Unlock()
+	if tm == nil {
```

Comment:
Since I can't think of a reason why a `nil` timer would be different from `timer does not exist` I think the `var, ok` pattern creates ambiguity since both `var == nil` and `ok == false` mean the timer does not exist. In a technical sense this doesn't matter but with that it depends on the style of the developer to choose a pattern which means we have to remember it and point it out during code reviews. 

It also creates ambiguity for us since you now cannot be sure that `nil` does **not** mean `timer does not exist`. Why else would we use that pattern? But what does it mean then? To find out you have to go back and read the code. 

The last thing is that this makes writing tests more complex since you now need to compare two fields for the `ok == true` case. 

These would be my arguments for not using the `ok bool`. I get that you need it when the zero value has meaning in a map but we don't have that since this is a data structure for a specific use case and not a generic one which has to cover every use case.

```go
# both valid and correct
if tm := m.Get("foo"); tm != nil { ... }
if tm, ok := m.Get("foo"); !ok { ... }

# this no longer works
if got, want := m.Get("foo"), tm; got != want {
	t.Fatalf("got %v want %v", got, want)
}

# becomes 
gottm, ok := m.Get("foo")
if !ok {
	t.Fatalf("got %v want true", ok)
}
if got, want := gottm, tm; got != want {
	t.Fatalf("got %v want %v", got, want)
}
```

Ah look, apparently I **do** care enough to argue my case :)

I'll change this in a separate commit so that we can see the difference.